src/import-shadow-hash.c

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "storage.h"
+#include "module.h"
+#include "entry.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <shadow.h>
+
+/**
+ * CreateEntryFromShadowHash:
+ * @username: Username containing the shadow hash.
+ * @sdhash: User's hashed shadow password.
+ */
+gboolean CreateEntryFromShadowHash(const char *username, char *sdhash) {
+  // Create a new Cache Entry.
+  CacheEntry *entry = CacheEntryNew();
+  GBytes* gbytes = g_bytes_new(sdhash, strlen(sdhash));
+  // Set the newly created entry hash using the /etc/shadow hash.
+  CacheEntryHashSet(entry, CACHE_ENTRY_ALGORITHM_CRYPT, gbytes, NULL);
+  g_bytes_unref(gbytes);
+  GError *error = NULL;
+  CacheStorage *storage = CacheStorageNew(DEFAULT_STORAGE_PATH);
+  if (!CacheStoragePutEntry(storage, username, entry, &error))
+      goto done;
+
+  done:
+    if (error) {
+      fprintf(stderr, "Failed to create Cache Entry.\n");
+      return FALSE;
+    } else {
+      fprintf(stdout, "Successfully created Cache Entry.\n");
+      return TRUE;
+    }
+}
+
+
+/**
+ * ReadShadowFile:
+ * @username: Username to read hashed password for.
+ */
+gboolean ReadShadowFile(const gchar *username, const gchar *path) {
```

Comment:
Move to util.h/util.c and test this in util_test.c.

Instead of doing something with the hash you get as a side-effect (calling CreateEntryFromShadowHash), return the hash let the caller create the entry.

The function signature would then be:
`GBytes *ReadShadowFile(const gchar *path, const gchar *username, GError **error)`

If there's an error, you return NULL and use g_set_error to set error. If there's no error, return a GBytes with the hash. That GBytes would be passed to CacheEntryHashSet().

In the test, you'll use `CacheTestGetDataPath("shadow")` to get the path to "testdata/shadow", input that into the path arg, and check the result for various users in your shadow file.